server.js

```diff
@@ -1,28 +1,133 @@
+    status: error.statusCode,
+    message: error.message,
+    success: false,
+  });
+};
+
+// Define a middleware to control the request method
+const methodController = (req, res, next) => {
+  if (req.method !== "GET") {
+    const err = new Error(
+      `${req.method} method is not implemented yet. Try GET method instead.`
+    );
+    err.statusCode = 501;
+    throw err;
+  }
+  next();
+};
+
+// Define a middleware to refuse all the query params
+const queryParamContoller = (req, res, next) => {
+  const queryParams = req.query;
+  if (Object.keys(queryParams).length > 0) throw new Error();
+  next();
+};
+
 // Add middlewares to enable cors and json body parsing
 app.use(cors());
 app.use(express.json());
+// Add middleware to enable fetching static files from public folder
+app.use(express.static("public"));
+// Add middleware to limit the request method to GET
+app.use(methodController);
+
+// Endpoints
+// Homepage endpoint: display API endpoints & instruction
+// middleware: if homepage gets any query params -> throw 400 error
+app.get("/", queryParamContoller, (req, res) => {
+  let endpoints = expressListEndpoints(app);
+  endpoints.map(
+    (endpoint, index) => (endpoint.description = endpointDescr[index])
+  );
+  res.render("index.ejs", { data: endpoints });
+});
+
+// songs endpoint: display all songs and filter songs based on genre and/or artist
+app.get("/songs", (req, res, next) => {
+  const queryParams = req.query;
+  const { genre, artist } = queryParams;
+  let songs = topMusicData;
+  try {
+    if (Object.keys(queryParams).length > 0) {
+      if (genre && artist) {
+        songs = topMusicData.filter(
+          song =>
+            song.genre === genre.toLowerCase() &&
+            song.artistName.toLowerCase() === artist.toLowerCase()
+        );
+      } else if (genre) {
+        songs = topMusicData.filter(song => song.genre === genre.toLowerCase());
+      } else if (artist) {
+        songs = topMusicData.filter(
+          song => song.artistName.toLowerCase() === artist.toLowerCase()
+        );
+      } else {
+        throw new Error();
+      }
+      if (songs.length === 0) {
+        const err = new Error(
+          `There are no songs matching the ${genre ? "genre:" + genre : ""}${
+            genre && artist ? " and " : ""
+          }${artist ? "artist:" + artist : ""}...`
+        );
+        err.statusCode = 404;
+        throw err;
+      }
+    }
+    res.json(songs);
+  } catch (error) {
+    next(error);
+  }
+});
+
+//song/id endpoint: display a single song
+// middleware: if homepage gets any query params -> throw 400 error
+app.get("/songs/:songId", queryParamContoller, (req, res, next) => {
+  const songId = req.params.songId;
+  const song = topMusicData.find(song => song.id === +songId);
+  if (song) {
+    res.json(song);
+  } else {
+    const err = new Error(
+      `There are no songs matching the id:${songId}... Make sure you are inquering id from 1 - 50.`
+    );
```

Comment:
Good that you're handling unavailable requests and I like that you are suggesting to the user how to do it correctly! Very user friendly 👍